File: hw/otavalo/u3/hw35StrategyPersistence/Srategy/StrategyJs/InstrumentsManagementJS/src/main/javascript/ec/edu/espe/instruments/utils/JsonPersistence.js
const fs = require("fs");
const path = require("path");
const Persistence = require("./Persistence");
const Instrument = require("../model/Instrument");

class JsonPersistence extends Persistence {
  static FILE_NAME = "Instruments.json";

  async create(instrument) {
    const instruments = await this.read();
    if (instruments.some(t => t.getId() === instrument.getId())) return false;
    instruments.push(instrument);
    return this._saveAll(instruments);
  }

  async read() {
    const filePath = path.resolve(JsonPersistence.FILE_NAME);
    if (!fs.existsSync(filePath)) return [];
    const raw = fs.readFileSync(filePath, "utf-8");
    if (!raw.trim()) return [];
    try {
      const arr = JSON.parse(raw);
      return (arr || []).map(o => new Instrument(o.id, o.name, o.price, o.materials || [], o.priceWithIva));
    } catch {
      return [];
    }
  }

  async update(id, instrument) {
    const instruments = await this.read();
    const idx = instruments.findIndex(t => t.getId() === id);
    if (idx === -1) return false;
    instruments[idx] = instrument;
    return this._saveAll(instruments);
  }

  async delete(id) {
    const instruments = await this.read();
    const filtered = instruments.filter(t => t.getId() !== id);
    if (filtered.length === instruments.length) return false;
    return this._saveAll(filtered);
  }

  async find(id) {
    const instruments = await this.read();
    return instruments.find(t => t.getId() === id) || null;
  }

  _saveAll(instruments) {
    try {
      const filePath = path.resolve(JsonPersistence.FILE_NAME);
      const plain = instruments.map(t => ({
        id: t.getId(),
        name: t.getName(),
        price: t.getPrice(),
        materials: t.getMaterials(),
        priceWithIva: t.getPriceWithIva()
      }));
      fs.writeFileSync(filePath, JSON.stringify(plain, null, 2), "utf-8");
      return true;
    } catch {
      return false;
    }
  }
}

module.exports = JsonPersistence;

File: hw/collaguazo/u3/hw35Strategy/StrategyPy/Instruments/src/main/python/ec/edu/espe/instruments/utils/JsonPersistence.py
import os
import json
from ec.edu.espe.instruments.utils.Persistence import Persistence
from ec.edu.espe.instruments.model.Instrument import Instrument

class JsonPersistence(Persistence):
    FILE_NAME = "instruments.json"

    def create(self, instrument):
        instruments = self.read()
        if any(t.getId() == instrument.getId() for t in instruments):
            return False
        instruments.append(instrument)
        return self._save_all(instruments)

    def read(self):
        if not os.path.exists(self.FILE_NAME):
            return []
        try:
            raw = open(self.FILE_NAME, "r", encoding="utf-8").read().strip()
            if not raw:
                return []
            arr = json.loads(raw)
            instruments = []
            for o in (arr or []):
                instruments.append(Instrument(o["id"], o["name"], float(o["price"]), o.get("materials", []) or [], float(o["priceWithIva"])))
            return instruments
        except:
            return []

    def update(self, id, instrument):
        instruments = self.read()
        for i, t in enumerate(instruments):
            if t.getId() == id:
                instruments[i] = instrument
                return self._save_all(instruments)
        return False

    def delete(self, id):
        instruments = self.read()
        new_instruments = [t for t in instruments if t.getId() != id]
        if len(new_instruments) == len(instruments):
            return False
        return self._save_all(new_instruments)

    def find(self, id):
        for t in self.read():
            if t.getId() == id:
                return t
        return None

    def _save_all(self, instruments):
        try:
            arr = [{
                "id": t.getId(),
                "name": t.getName(),
                "price": t.getPrice(),
                "materials": t.getMaterials() or [],
                "priceWithIva": t.getPriceWithIva()
            } for t in instruments]
            with open(self.FILE_NAME, "w", encoding="utf-8") as f:
                json.dump(arr, f, indent=2)
            return True
        except:
            return False
